### ISSUE TITLE
Restoring fails with TargetInvocationException

### ISSUE THREAD

**Issue by Phaiax** [duplicati/duplicati#3542]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.4.8_canary_2018-12-13
- **Operating system**: Windows 10 (Version 10.0.17134 Build 17134)
- **Backend**: SSH@Synology

## Description

I tried restoring one file, to a special destination folder, "Write Multiple versions with timestamp" was selected, then I clicked "restore".
The Website stated "Dateien und Ordner erfolgreich wiederhergestellt." (="Files and folders were successfully restored"), at the same time a Windows notification and a Website notification appeared, stating "Error while running Foo-Bar-Backup" and "Ein Aufrufziel hat einen Ausnahmefehler verursacht." (="A target invocation has produced an exceptional error"). Clicking "Show" on the error forwarded to the backup log which did not contain any information about the error. Instead in the system log under "About", I found the backtrace below.
In the restore destination folder, no files appeared (still empty).

## Steps to reproduce

1. Install Duplicaty, the current newest canary
2. Add backup, no special options, to SSH, encrypted
3. Start backup (on the remote location, a few `duplicati-...-*.zip.aes` files appeared.
4. Try restoring a file

5. I tried restarting the dupicati server, but the problem persists.

- **Actual result**: Restoring did not work, even if the website stated it was successful (but there were error notifications)
- **Expected result**: Restoring should work or the website should also state an error. The "Show" link in the notification opens the correct log.

If I can help you any further, please ask.
Thank you for your time.
Daniel

## Debug log

The text in the stacktrace can be translated to:
"Empty passphrase not allowed; Parameter name: passphrase at Duplicati.Library.Encryption.AESEncryption..ctor("

```
System.Reflection.TargetInvocationException: Ein Aufrufziel hat einen Ausnahmefehler verursacht. 
---> System.ArgumentException: Leere Passphrase nicht erlaubt Parametername: passphrase bei String passphrase, Dictionary`2 options) 
--- Ende der internen Ausnahmestapelüberwachung --- 
bei System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) 
bei System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
bei System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark) 
bei System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes) 
bei System.Activator.CreateInstance(Type type, Object[] args) 
bei Duplicati.Library.DynamicLoader.EncryptionLoader.EncryptionLoaderSub.GetModule(String fileExtension, String passphrase, Dictionary`2 options) 
bei Duplicati.Library.Main.BackendManager..ctor(String backendurl, Options options, IBackendWriter statwriter, LocalDatabase database) 
bei Duplicati.Library.Main.Operation.RestoreHandler.DoRun(LocalDatabase dbparent, IFilter filter, RestoreResults result) 
bei Duplicati.Library.Main.Operation.RestoreHandler.Run(String[] paths, IFilter filter) 
bei Duplicati.Library.Main.Controller.<>c__DisplayClass14_0.<Restore>b__0(RestoreResults result) 
bei Duplicati.Library.Main.Controller.RunAction[T](T result, String[]& paths, IFilter& filter, Action`1 method) 
bei Duplicati.Library.Main.Controller.Restore(String[] paths, IFilter filter) 
bei Duplicati.Server.Runner.Run(IRunnerData data, Boolean fromQueue)
```

**Comment by NerdGuruDevelopers** [comment]
Having an identical error:

`System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ArgumentException: Empty passphrase not allowed
Parameter name: passphrase
  at Duplicati.Library.Encryption.AESEncryption..ctor (System.String passphrase, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x00013] in <38f40f254bb94cb3afc644103e1e7581>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x00005] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x00017] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0007a] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Reflection.MonoCMethod.Invoke (System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x0022b] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00095] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at Duplicati.Library.DynamicLoader.EncryptionLoader+EncryptionLoaderSub.GetModule (System.String fileExtension, System.String passphrase, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x00049] in <43fe28e9c2764d0d8f458f05671e026c>:0 
  at Duplicati.Library.DynamicLoader.EncryptionLoader.GetModule (System.String fileextension, System.String passphrase, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x00000] in <43fe28e9c2764d0d8f458f05671e026c>:0 
  at Duplicati.Library.Main.BackendManager..ctor (System.String backendurl, Duplicati.Library.Main.Options options, Duplicati.Library.Main.IBackendWriter statwriter, Duplicati.Library.Main.Database.LocalDatabase database) [0x000d8] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Library.Main.Operation.RestoreHandler.DoRun (Duplicati.Library.Main.Database.LocalDatabase dbparent, Duplicati.Library.Utility.IFilter filter, Duplicati.Library.Main.RestoreResults result) [0x00019] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Library.Main.Operation.RestoreHandler.Run (System.String[] paths, Duplicati.Library.Utility.IFilter filter) [0x00062] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Library.Main.Controller+<>c__DisplayClass14_0.<Restore>b__0 (Duplicati.Library.Main.RestoreResults result) [0x0001c] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Library.Main.Controller.RunAction[T] (T result, System.String[]& paths, Duplicati.Library.Utility.IFilter& filter, System.Action`1[T] method) [0x00271] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Library.Main.Controller.Restore (System.String[] paths, Duplicati.Library.Utility.IFilter filter) [0x00021] in <d13d696d40bb4d4da88c121875e81b80>:0 
  at Duplicati.Server.Runner.Run (Duplicati.Server.Runner+IRunnerData data, System.Boolean fromQueue) [0x003b8] in <17472153ff7746f0ac5b06da5a6e3c09>:0 `

**Comment by NickPyz** [comment]
+1

Duplicati 2.0.4.9 Canary
Windows 10 Pro (64) ver. 1803

All of my backup jobs are AES encrypted by Duplicati and backed-up to various remote back-ends - SSH, AWS S3, Backblaze B2, OneDrive.  None of them can be restored. I get the same error: "Restoring fails with Target Invocation Exception".

Critical error, IMO.

**Comment by warwickmm** [comment]
I was able to reproduce this.  The fix in pull request #3563 resolved the issue for me.
